lldpd: allow glob patterns for interface(s) parsing

systemcrash opened the pull request:
For interface related parameters, the man page documents functionality such as:

```
*,!eth*,!!eth1

uses all interfaces, except interfaces starting with "eth",
but including "eth1".
```

Before this change: only interfaces like `eth0, eth1` etc were accepted. 
After this change: interface masks and inversions are now also accepted: `*,!eth*,!!eth1,lan1,lan2,lan3,!lan4`

Tested extensively on: 22.03.6

'Optionally' depends upon #14850 (from which it optionally inherits a package release bump)

Review welcome: @robimarko @stintel @pprindeville

pprindeville commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -81,9 +81,25 @@ get_config_cid_ifaces() {
+		if network_get_device _l3dev "$_suffix" || [ -e "/sys/class/net/$_suffix" ]; then
+			if [ -n "$_prefix" ]; then
+				# prepend the stripped "!" or "!!" prefix here if we had one
+				append _ifnames "$_prefix${_l3dev:-$_suffix}" ","
```
Am I missing something?  Isn't line 92 all you need instead of 90-95?  If `$_prefix` is empty then it will harmless expand to the empty string...

systemcrash replied:
Absolutely right.

pprindeville commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -81,9 +81,25 @@ get_config_cid_ifaces() {
+			fi
+		else
+			case $_iface in 
+				*"*"*)
```
Is the third `*` necessary?  I think you only want a `*` suffix for globbing, don't you?

systemcrash replied:
I'm not sure - consider the patterns: `!*:*`, `!eth*` or even `eth*1`.

pprindeville replied:
okay. as long as `lldpd` itself can handle those patterns.

systemcrash replied:
OK. Yeah, that's the raison d'être for this PR. lldpd handles those.

pprindeville reviewed (commented):
Some minor questions...

pprindeville reviewed (approved):
LGTM

systemcrash commented:
@Npeca75 good if you could test this

pprindeville commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -188,6 +198,8 @@ write_lldpd_conf()
 	[ -n "$lldp_platform" ] && echo "configure system platform" "\"$lldp_platform\"" >> "$LLDPD_CONF"
 	[ -n "$lldp_tx_interval" ] && echo "configure lldp tx-interval $lldp_tx_interval" >> "$LLDPD_CONF"
 	[ "$lldp_tx_hold" -gt 0 ] && echo "configure lldp tx-hold $lldp_tx_hold" >> "$LLDPD_CONF"
+	[ "$lldp_capadv" -gt 0 ] && echo "configure lldp capabilities-advertisements" >> "$LLDPD_CONF"
```
I think as a bool that `-ne 0` would be more readable/intuitive.

systemcrash replied:
Valid. As long as both work with potentially non-boolean values. Some settings (elsewhere) have been co-opted to later become ternary or greater choice settings.

systemcrash replied:
Here's the thing: the rest of the file uses `-gt 0` so it seems like a refactor is in order, just so these two `-ne 0` don't raise questions later of "why did they do that?"

pprindeville replied:
Sorry, I thought the others were for integers and not booleans.

In any case, not crucial.

pprindeville reviewed (changes requested):
Version bump on `PKG_RELEASE`?

pprindeville commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -349,34 +375,36 @@ reload_service() {
 
 stop_service() {
-	rm -rf ${LLDPD_RUN} "$LLDPSOCKET" 2> /dev/null
+	rm -rf ${LLDPD_RUN} "$LLDPSOCKET" >/dev/null 2>&1
```
What message is this going to generate such that `stderr` needs to be discarded?

howels replied:
It's possible to build lldpd without socket support. References to the socket are discarding errors throughout the script, presumably to allow for builds without the socket support.

systemcrash replied:
Or, in case no service was actually running when someone runs 'stop'.

pprindeville replied:
That's fine, but removing a non-existent file with `rm -rf` won't generate any output anyway...

```
       -f, --force
              ignore nonexistent files and arguments, never prompt
```

systemcrash replied:
That's true. The spurious output at the moment needs fixing, however, so the parameter order must be changed.

systemcrash commented:
Kind reminder to merge this patch-set necessary for lldpd and which holds up the luci implementation which has been waiting (too long) for merge in openwrt/luci#6456

@Ansuel @neheb @jow- @robimarko @stintel @openwrt-bot @dangowrt @aparcar @Noltari @hauke @1715173329

Ansuel commented:
@systemcrash i cherry-pick 2 commit from here just to start the ball rolling. I think also the configuration commits can be merged safely?

Ansuel reviewed (commented):
https://github.com/openwrt/openwrt/pull/14867/commits/819c328798bc18005edcd6a1ab58d70c146b8014 I'm not following where this is actually used in later commits...

systemcrash commented:
> @systemcrash i cherry-pick 2 commit from here just to start the ball rolling. I think also the configuration commits can be merged safely?

Yes.

Everything here has been tested with various lldpd versions on at least 22 and 23. Backwards behavior (except the restart messages) is maintained. The interface parsing is necessary - see first message.

systemcrash commented:
> I'm not following where this is actually used in later commits...

In reference to what?

Ansuel commented:
> > I'm not following where this is actually used in later commits...
> 
> In reference to what?

I updated the message, just github being confusing with reviewing single commits.

Ansuel commented:
@systemcrash for the first commit, can you give me an example of interface/pattern that makes the thing explode? I would love to find a way to skip setting the noglob... Maybe we are missing something and we should just use "" ?

systemcrash commented:
> [819c328](https://github.com/openwrt/openwrt/commit/819c328798bc18005edcd6a1ab58d70c146b8014) I'm not following where this is actually used in later commits...

That whole commit is self contained. The function now describes what it does. And it's now used by both of the "interface" handling settings which get ips for management.

systemcrash commented:
> @systemcrash for the first commit, can you give me an example of interface/pattern that makes the thing explode? I would love to find a way to skip setting the noglob... Maybe we are missing something and we should just use "" ?

If you use any interface name that contains "*", the init script starts to process files in its directory(!) instead. The noglob is only on fir a moment. 

What is the concern here? The noglob is only actual under that shell instance.

Ansuel commented:
> That whole commit is self contained. The function now describes what it does. And it's now used by both of the "interface" handling settings which get ips for management.

yes but it's not clear to me why... It's just a rename or a cleanup but the variant function (where a different config or arg can be used) is not used anywhere in later commit... It feels to me this is something related to other changes that will come later???

Honestly I would just rename the function (and keep the 'config' part hardcoded) or even drop the pr from here and propose it when the change is actually required?

(not saying it's bad... just saying that i'm confused on the why part)

Ansuel commented:
> What is the concern here? The noglob is only actual under that shell instance. Have you encountered undesirable behavior using this?

just taste and searching for better solution that using the set function... Soooo an interface like eth* should trigger the problem?

I want to repro it on a running system so I can waste some time finding a solution.

systemcrash commented:
> > What is the concern here? The noglob is only actual under that shell instance. Have you encountered undesirable behavior using this?
> 
> just taste and searching for better solution that using the set function... Soooo an interface like eth* should trigger the problem?
> 
> I want to repro it on a running system so I can waste some time finding a solution.

hahahah. Well, this commit must be used (just comment out the noglob part) to ensure you see the problem. And set an interface like e.g. `*,!eth2` (meaning: use all interfaces except eth2)

jow- commented:
Changing the interface to an actual UCI list type and using `config_list_foreach()` would be one solution as it would avoid the need for a `for ... in ...` statement. It would also be closer to the spirit of uci. I still think that we should avoid adding new "micro formats" into uci. Add one config list "interfaces" and two other lists "exclude_interfaces" and "include_interfaces", then build that `x,!y,!!z` list from those.

systemcrash commented:
> > That whole commit is self contained. The function now describes what it does. And it's now used by both of the "interface" handling settings which get ips for management.
> 
> yes but it's not clear to me why... It's just a rename or a cleanup but the variant function (where a different config or arg can be used) is not used anywhere in later commit... It feels to me this is something related to other changes that will come later???

I see. No - it just looks better as a variant. 

I have in mind TLV handling: There's #17872 for that. It definitely needs variant format to manage interfaces, for which TLV can be independently assigned.

Ansuel commented:
> I have in mind TLV handling: There's #17872 for that. It definitely needs variant format to manage interfaces, for which TLV can be independently assigned.

ok then please I would prefer if that commit moves to #17872 if it's not a problem for you

Ansuel commented:
> Changing the interface to an actual UCI list type and using `config_list_foreach()` would be one solution as it would avoid the need for a `for ... in ...` statement. It would also be closer to the spirit of uci. I still think that we should avoid adding new "micro formats" into uci. Add one config list "interfaces" and two other lists "exclude_interfaces" and "include_interfaces", then build that `x,!y,!!z` list from those.

@systemcrash considering the luci app is still WIP and the glob wasn't a thing previously... Is it possible to implement this instead?

systemcrash commented:
> > I have in mind TLV handling: There's #17872 for that. It definitely needs variant format to manage interfaces, for which TLV can be independently assigned.
> 
> ok then please I would prefer if that commit moves to #17872 if it's not a problem for you

Sure. No problem - as long as the opposite discussion does not happen on the TLV commit. 😌

Ansuel commented:
> Sure. No problem - as long as the opposite discussion does not happen on the TLV commit. 😌

nha on the TLV commit it will be actually used so it's usage will be more clear...

I also merged the other 2 small commit... reducing this pr to just 3 commits

systemcrash commented:
@jow- 
> Changing the interface to an actual UCI list type and using `config_list_foreach()` would be one solution as it would avoid the need for a `for ... in ...` statement. It would also be closer to the spirit of uci. I still think that we should avoid adding new "micro formats" into uci. Add one config list "interfaces" and two other lists "exclude_interfaces" and "include_interfaces", then build that `x,!y,!!z` list from those.

Seems reasonable. It's been more than a month since I did this, so reasoning for decisions is not front and centre in my memory any longer.

I spent a lot of time ensuring I didn't go around changing (too much) of the way things work today. And the luci part is a done deal - I wrote GUI handlers to pack interfaces into a CSV because error checking a DynamicList was like dogshit: you cannot easily check what you've burnt into the HTML there - but if it's just a (custom)List, `.validate` is really simple, then `.write` becomes 

`			return this.super('write', [ section_id, value.join(',') ]);`


Having it as @jow- suggests might work under the hood. Such a change won't disturb existing configs?

Ansuel commented:
> Such a change won't disturb existing configs?

what are the current accepted values for interface ?

systemcrash commented:
> > Such a change won't disturb existing configs?
> 
> what are the current accepted values for 

@Ansuel 
An interface name that exists: as a L3device or not at all

systemcrash commented:
Emphasis on the singular

jow- commented:
Just noticed that the implementation of the interface pattern parsing is also incomplete: lldpd uses `fnmatch()` under the hood to test the glob patterns, so a valid pattern might also contain `?` or a `[...]` character class (potentially containing a `!`, so the `_suffix=${_iface##*"!"}` expression would incorrectly split a given pattern in the middle I believe.

Ansuel commented:
@jow- supporting all of that doesn't seems trivial to me... unless sed or grep can be used... guess your idea of include and exclude is much simple to implement.

systemcrash commented:
> Just noticed that the implementation of the interface pattern parsing is also incomplete: lldpd uses `fnmatch()` under the hood to test the glob patterns, so a valid pattern might also contain `?` or a `[...]` character class (potentially containing a `!`, so the `_suffix=${_iface##*"!"}` expression would incorrectly split a given pattern in the middle I believe.

Interesting - I don't want to say "unlikely" but those variants are not listed in the man pages, so I figured one would not encounter those. The `*` and `!` seem like enough. 

At that point though, we are no longer dealing with a prefix or suffix, and the `network_get_device` part will fail anyway, I think.

jow- commented:
> At that point though, we are no longer dealing with a prefix or suffix, and the network_get_device part will fail anyway, I think.

Your code would silently discard all valid patterns that neither match an actual device name nor contain a `*`, so a valid pattern like `eth[0-9]` or `sw?p?` would not end up in the generated config.

systemcrash commented:
> > At that point though, we are no longer dealing with a prefix or suffix, and the network_get_device part will fail anyway, I think.
> 
> Your code would silently discard all valid patterns that neither match an actual device name nor contain a `*`, so a valid pattern like `eth[0-9]` or `sw?p?` would not end up in the generated config.

I strove to follow documented bahviour in the man pages, and expose only that in the GUI. But if my code discards it, so does the existing code (maintaining behaviour). 

The include/exclude seems like a robust approach, however.

Ansuel commented:
@systemcrash what I want to have it clear is if the current supported behavior is for a single interface (or a list of) only or have some kind of support for the pattern... aka the * wasn't supported at all.

systemcrash commented:
> Your code would silently discard all valid patterns that neither match an actual device name nor contain a `*`, so a valid pattern like `eth[0-9]` or `sw?p?` would not end up in the generated config.

But again - how do you envisage an include/exclude deal with these patterns? Doing an exhaustive search under `/sys`?

jow- commented:
As for backward compatibility:

Doing a `config_get` on a uci list value will return a space separated string, as if this value was declared as `option foo ...` instead of `list foo ...`. Doing a `config_list_foreach` on an `option ...` value will only return a single item though, even if that happens to contain spaces.

By querying the `_LENGTH` property, one can support both variants:
https://github.com/openwrt/openwrt/blob/main/package/network/services/dnsmasq/files/dnsmasq.init#L685-L699

systemcrash commented:
> @systemcrash what I want to have it clear is if the current supported behavior is for a single interface (or a list of) only or have some kind of support for the pattern... aka the * wasn't supported at all.

(It was just long enough ago I started to forget. ) Current behaviour supports a CSV, I think. But only of devices which exist.

Globs started the 'files in directory' problem.

jow- commented:
But again - how do you envisage an include/exclude deal with these patterns? Doing an exhaustive search under /sys?

Attempt to resolve the given value as l3device of an ifname or pass as-is. I would make no attempt to resolve `foo*` into anything.

systemcrash commented:
> > But again - how do you envisage an include/exclude deal with these patterns? Doing an exhaustive search under /sys?
> 
> Attempt to resolve the given value as l3device of an ifname or pass as-is. I would make no attempt to resolve `foo*` into anything.

OK - so... we drop prefix/suffix (as I wrote) and.. for example, if `network_get_device` fails, just take the pattern verbatim, assuming it's what the user wants?

jow- commented:
> OK - so... we drop prefix/suffix (as I wrote) and.. for example, if network_get_device fails, just take the pattern verbatim, assuming it's what the user wants?

That would mirror the existing behavior, yes. A value of `lan` would likely become `br-lan`. A value of `br-lan` or `lan*` would remain `br-lan` or `lan*` respectively.

systemcrash commented:
> Which brings me to another point (a problem that predated your changes)... we don't actually want to resolve the l3dev of a logical interfdace but the l2dev (`network_get_physdev`).

I was wondering about this bit. Because lldpd is interested in ... the link layer.

jow- commented:
The attempt-to-resolve-value-to-netdev logic could be confined to the `list interface ...` uci option only, while `list include_interface ...` and `list exclude_interface ...` would forward their values as-is (while prepending `!` and `!!` respectively). This way a user could enter `wan` in the `include_interfaces` list and it should always get forwarded.

jow- commented:
Regex != glob. https://man7.org/linux/man-pages/man7/glob.7.html (ctrl-f "Complementation"). 

> An expression "[!...]" matches a single character, namely any
> character that is not matched by the expression obtained by
> removing the first '!' from it.  (Thus, "[!]a-]" matches any
> single character except ']', 'a', and '-'.)

systemcrash commented:
> Regex != glob. https://man7.org/linux/man-pages/man7/glob.7.html (ctrl-f "Complementation").
> 
> > An expression "[!...]" matches a single character, namely any
> > character that is not matched by the expression obtained by
> > removing the first '!' from it.  (Thus, "[!]a-]" matches any
> > single character except ']', 'a', and '-'.)

If that's a valid glob, it should work in the ash shell directly from the command line, right?

jow- commented:
Here's an example to illustrate it:

```
jow@j7:~$ cat /tmp/fnmatch.c 
#include <stdio.h>
#include <fnmatch.h>

int main(int argc, char **argv) {
	const char *pattern = "eth[!89]";
	const char *values[] = {
		"eth0",
		"eth8",
		"ethX"
	};

	for (size_t i = 0; i < sizeof(values)/sizeof(values[0]); i++) {
		printf("%s matches %s? %d\n",
			pattern, values[i],
			fnmatch(pattern, values[i], 0) != FNM_NOMATCH);
	}

	return 0;
}
jow@j7:~$ gcc -o /tmp/fnmatch /tmp/fnmatch.c && /tmp/fnmatch
eth[!89] matches eth0? 1
eth[!89] matches eth8? 0
eth[!89] matches ethX? 1
jow@j7:~$ 
```

The lldpd daemon uses the same fnmatch() function for its glob processing:
https://github.com/lldpd/lldpd/blob/master/src/daemon/pattern.c#L59

systemcrash commented:
> Same for shell:
> 
> ```
> jow@j7:~$ mkdir /tmp/glob
> jow@j7:~$ cd /tmp/glob/
> jow@j7:/tmp/glob$ touch eth0 eth8 ethX
> jow@j7:/tmp/glob$ echo eth[!89]
> eth0 ethX
> jow@j7:/tmp/glob$ 
> ```

Curious - I could not get a simple `ls` to use it. Also cannot seem to reproduce that. 🙈

jow- commented:
```
jow@j7:~$ ssh root@fb7530


BusyBox v1.36.1 (2023-11-14 13:38:11 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 23.05.2, r23630-842932a63d
 -----------------------------------------------------
root@fb7530:~# mkdir /tmp/glob
root@fb7530:~# cd /tmp/glob/
root@fb7530:/tmp/glob# touch eth0 eth8 ethX
root@fb7530:/tmp/glob# echo eth[!89]
eth0 ethX
root@fb7530:/tmp/glob# ls -l eth[!89]
-rw-r--r--    1 root     root             0 Apr 12 13:02 eth0
-rw-r--r--    1 root     root             0 Apr 12 13:02 ethX
root@fb7530:/tmp/glob# 
```

systemcrash commented:
That's where I'm testing ;) but it does not work as a prefix

> ```
> root@fb7530:~# mkdir /tmp/glob
> root@fb7530:~# cd /tmp/glob/
> root@fb7530:/tmp/glob# touch eth0 eth8 ethX
> root@fb7530:/tmp/glob# echo eth[!89]
> eth0 ethX
> root@fb7530:/tmp/glob# ls -l eth[!89]
> -rw-r--r--    1 root     root             0 Apr 12 13:02 eth0
> -rw-r--r--    1 root     root             0 Apr 12 13:02 ethX
> root@fb7530:/tmp/glob# 
> ```

```
root@asdf:/tmp# touch tboard.json
root@asdf:/tmp# ls *.json
board.json                     luci-indexcache.bd5ea4fa.json  tboard.json
root@asdf:/tmp# echo [!t]board.json
[!t]board.json
```

jow- commented:
Try touch `aboard.json` and repeat your test.

systemcrash commented:
> It does, but `[!t]` consumes exactly one character (like `?`), you don't have any `Xboard.json` where `X` is any character except `t`.
> 
> Try `touch aboard.json` and repeat your test.

OK yeah - that worked now. Seems logical... although takes a while to understand it as logical.

Ansuel commented:
@systemcrash btw the noglob can be handled by just spamming "" on variables and prevent shell expanding it. Can you test that? On my little snippet it works.

systemcrash commented:
I pushed the least invasive version of changes (globs are just `else` case) while retaining current behaviour.

@Ansuel I'll go test your suggestion now.

systemcrash commented:
@Ansuel can you show me your snippet please?

systemcrash commented:
The current version still need noglob. 
```sh
get_config_cid_ifaces() {
	# Set noglob to prevent '*' capturing diverse file names
	# set -o noglob
	local _ifaces
	config_get _ifaces 'config' "$2"

	local _iface _ifnames=""
	for _iface in $_ifaces; do

		local _l2device=""
		if network_get_physdev _l2device "$_iface" || [ -e "/sys/class/net/$_iface" ]; then
			append _ifnames "${_l2device:-$_iface}" ","
		else
			# Glob case (interface is e.g. '!eth1' or 'eth*' or '*')
			append _ifnames "$_iface" ","
		fi
	done

	export -n "${1}=$_ifnames"
	# set +o noglob
}
```

Ansuel commented:
@systemcrash what is the interface value used?

systemcrash commented:
> @systemcrash what is the interface value used?

As yours: "*"

Ansuel commented:
@systemcrash you didn't escape the $_ifaces in the for loop that is what is getting expanded.

systemcrash commented:
> @systemcrash you didn't escape the $_ifaces in the for loop that is what is getting expanded. you need to use "$_ifaces"

shellcheck complains: `Since you double quoted this, it will not word split, and the loop will only run once.`

I still get globbed. :)

systemcrash commented:
If the config has more than one entry there, e.g. "*,!eth1" then the loop does not work. 

Well, IMHO `noglob` is viable until some nice sunny day in the future when include and exclude list handling can start occupying more script space. It's my understanding that noglob is only active in the current namespace (shell invocation). Or is it somehow otherwise and I've just not noticed it in openwrt ash?

The idea is that users just could follow the lldpd man page syntax and use "*,!eth1,!eth2" in a single line (copy+paste).

systemcrash committed:
lldpd: extended interface(s) parsing to handle patterns

For interface type parameters, the man page documents patterns:
```
*,!eth*,!!eth1

uses all interfaces, except interfaces starting with "eth",
but including "eth1".
```

* Renamed `_ifname` to `_l2dev`.
* get the l2dev via network_get_physdev (and not l3dev)
* Glob pattern `*` is also valid - use noglob for this

The net result is that now interface 'names' including globs '*' and '!'
inversions are included in the generated lldpd configs.

Temporarily `set -o noglob` and then `set +o noglob` to disable & enable
globbing respectively, because when we pass `*` as an interface choice,
other file and pathnames get sucked in from where the init script runs,
and the `*` never makes it to lldpd.

Tested extensively on: 22.03.6, 23.05.3

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
[ squash with commit bumping release version ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

systemcrash commented:
> just taste and searching for better solution than using the set function...

@Ansuel  so.. `set` is in use all over the code-base. And `noglob` is namespace only. I don't like making each setting which uses an interface value in triplicate just to avoid using `noglob`. (Each TLV takes a `ports` parameter so three fields there also? 😦). noglob is for exactly this case.

I see some people have written about noglob "that could be bad blah blah because you can't do `ls *` any more" but they're only ever talking about turning glob off... and *leaving it off*.

Is this the only remaining objection in the last commit?

systemcrash commented:
@Ansuel be great if this can get a current review (using `noglob`). The luci part is just hanging there waiting for this.
